ledger-core/v2/virtual/request/vstaterequest.go

```diff
@@ -0,0 +1,170 @@
+	s := sm.(*SMVStateRequest)
+
+	injector.MustInject(&s.pulseSlot)
+	injector.MustInject(&s.messageSender)
+}
+
+func (s *SMVStateRequest) GetStateMachineDeclaration() smachine.StateMachineDeclaration {
+	return dSMVStateRequestInstance
+}
+
+func (s *SMVStateRequest) Init(ctx smachine.InitializationContext) smachine.StateUpdate {
+	return ctx.Jump(s.stepProcess)
+}
+
+func (s *SMVStateRequest) stepProcess(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	var (
+		objectCatalog = object.Catalog{}
+	)
+	objectSharedState, stateFound := objectCatalog.TryGet(ctx, s.Payload.Callee)
+	if !stateFound {
+		s.failReason = payload.Missing
+		return ctx.Jump(s.stepReturnStateUnavailable)
+	}
+
+	var (
+		semaphoreReadyToWork smachine.SyncLink
+	)
+
+	action := func(state *object.SharedState) {
+		semaphoreReadyToWork = state.ReadyToWork
+
+		descriptor := state.Descriptor()
+
+		if descriptor == nil {
+			panic(throw.IllegalState())
+		}
+
+		stateRef := descriptor.HeadRef()
+
+		immutableCounts, _ := state.ImmutableExecute.GetCounts()
+		mutableCounts, _ := state.MutableExecute.GetCounts()
+		s.objectStateReport = &payload.VStateReport{
+			AsOf:                  s.Payload.AsOf,
+			Callee:                s.Payload.Callee,
+			ImmutablePendingCount: int32(immutableCounts),
+			MutablePendingCount:   int32(mutableCounts),
+			LatestDirtyState:      *stateRef,
+		}
+
+		if s.Payload.RequestedContent^payload.RequestLatestDirtyState == 0 {
+			parent := descriptor.Parent()
+			proto, _ := descriptor.Prototype()
+
+			memory := descriptor.Memory()
+
+			s.objectStateReport.ProvidedContent = &payload.VStateReport_ProvidedContentBody{
+				LatestDirtyState: &payload.ObjectState{
+					Reference:   *stateRef,
+					Parent:      *parent,
+					State:       memory,
+					Deactivated: state.Deactivated,
+				},
+			}
+			if proto != nil {
+				s.objectStateReport.ProvidedContent.LatestDirtyState.Prototype = *proto
+			}
+		}
+	}
+
+	switch objectSharedState.Prepare(action).TryUse(ctx).GetDecision() {
+	case smachine.NotPassed:
+		return ctx.WaitShared(objectSharedState.SharedDataLink).ThenRepeat()
+	case smachine.Impossible:
+		ctx.Log().Fatal("failed to get object state: already dead")
+	case smachine.Passed:
+		// go further
+	default:
+		panic(throw.NotImplemented())
+	}
+
+	if ctx.AcquireForThisStep(semaphoreReadyToWork).IsNotPassed() {
+		return ctx.Sleep().ThenRepeat()
+	}
+
+	return ctx.Jump(s.stepSendResult)
+}
+
+func (s *SMVStateRequest) stepReturnStateUnavailable(ctx smachine.ExecutionContext) smachine.StateUpdate {
+	msg := &payload.VStateUnavailable{
+		Reason:   s.failReason,
+		Lifeline: s.Payload.Callee,
+	}
+
+	target := s.Meta.Sender
+
+	goCtx := ctx.GetContext()
+	s.messageSender.PrepareNotify(ctx, func(svc messagesender.Service) {
+		err := svc.SendTarget(goCtx, msg, target)
+		if err != nil {
+			ctx.Error(err)
```

Comment:
In description of this function is written that it's for termination SM.
Might be it's better to use ctx.Log().Error() here?
But I'm not sure